src/lib/formValidator.ts

```diff
@@ -0,0 +1,66 @@
+import {
+  IFieldValidator,
+  IFieldValidationResponse,
+  fieldValidatorLookup,
+} from "./fieldValidators";
+
+export interface IFormData {
+  [key: string]: string;
+}
+
+export interface IFieldValidatorLookup {
+  [key: string]: IFieldValidator;
+}
+
+export interface IFormValidationResponse {
+  [key: string]: IFieldValidationResponse;
+}
+
+export interface IFormError {
+  fieldId: string;
+  errors: string[];
+}
+
+export class FormValidator {
+  public static validate(
+    formData: IFormData,
+    validatorLookup: IFieldValidatorLookup = fieldValidatorLookup
+  ): IFormValidationResponse {
+    const fields = Object.entries(formData);
+
+    return fields.reduce((validatorResponse, [fieldId, value]) => {
+      validatorResponse[fieldId] = {
+        value: value,
+        ...validatorLookup[fieldId].validate(value),
+      };
+
+      return validatorResponse;
+    }, {});
```

Comment:
I <3 reducers/accumulators but if this isn't readable I can switch it out for something else.

- Array.prototype.reduce [docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce) and [Ruby equivalent](https://ruby-doc.org/core-2.4.0/Enumerable.html#method-i-reduce).
- [Precis on reducers in general](https://medium.com/async-la/a-short-and-sour-guide-to-reducers-b5b54d3bb018)

(Sorry if teaching to suck eggs)